Add cache for Socket.LocalEndPoint

CarnaViire opened the pull request:
Added _localEndPoint field to cache LocalEndPoint value. ~~Introduced RightEndPoint property to encapsulate clearing _localEndPoint on _rightEndPoint change.~~ ~~Cached value is cleared on connect, error and disconnect.~~ Cached value is cleared on error and disconnect. If the value is a wildcard address, it is also cleared on connect and accept.

Fix #1482

CarnaViire committed:
Add cache for LocalEndPoint

CarnaViire committed:
Add RightEndPoint property, clear _localEndPoint in RightEndPoint setter

CarnaViire committed:
Merge remote-tracking branch 'upstream/master' into issue-1482

CarnaViire commented:
/azp run runtime-libraries outerloop

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -315,32 +325,36 @@ public int Available
                 {
-                    return null;
-                }
+                    if (RightEndPoint == null)
+                    {
+                        return null;
+                    }
```
I think this is better outside the `if (_localEndPoint == null)` block.

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -23,9 +23,19 @@ public partial class Socket : IDisposable
+            private set
+            {
+                _rightEndPoint = value;
+                _localEndPoint = null;
```
I realized this might be an unnecessary step in many cases. Which tests do fail when you remove this line?

CarnaViire replied:
So actually none of them do 🤔 it seems strange and maybe we are lacking some tests? I am thinking that maybe we should clear `_localEndPoint` if new `_rightEndPoint` assigned is of different type than the old `_rightEndPoint` . However I'm not sure if this is a possible scenario? If it is, we should create a test for that.

CarnaViire replied:
I couldn't find any case in which original LocalEndPoint (without cache) would start to return a different value. Removed clearing _localEndPoint.

CarnaViire committed:
Merge remote-tracking branch 'upstream/master' into issue-1482

CarnaViire committed:
Merge remote-tracking branch 'upstream/master' into issue-1482

CarnaViire committed:
Merge remote-tracking branch 'upstream/master' into issue-1482

CarnaViire committed:
Roll back property changes

CarnaViire committed:
Fix _nonBlockingConnectRightEndPoint

CarnaViire commented:
/azp run runtime-libraries outerloop

karelz reviewed (approved):
LGTM, but I let @antonfirsov make the final ok ;)

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -325,23 +326,27 @@ public int Available
+                            }
                         }
                     }
+                    _localEndPoint = _rightEndPoint.Create(socketAddress);
```
According to [this discussion](https://github.com/dotnet/runtime/pull/1858#discussion_r368018995) the silent catch is most likely a leftover code we actually don't need.

antonfirsov commented:
I'm concerned if we should maybe reset `_localEndpoint` to `null` in methods where we also reset `_rightEndpoint`:

https://github.com/dotnet/runtime/blob/f8e92ecfdd329d1f0c4a1ccc1ae494282d9a3050/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L2685-L2712

If I'm not missing anything there will be no user callback querying `LocalEndpoint` in case of failure, however we may end up adding internal code doing so in the future, for example in a tracing call like this:
https://github.com/dotnet/runtime/blob/f8e92ecfdd329d1f0c4a1ccc1ae494282d9a3050/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L1083

Such a call may enforce caching a wrong value for `LocalEndpoint` permanently. I wonder if it makes sense to add tests to ensure such failing async methods will not change the sockets internal state in an unwanted way.

@tmds maybe you have some thoughts on this?

CarnaViire commented:
@antonfirsov Here is a list of all changes of _rightEndPoint that I have found in the code (sorry for the long list, I would put it under cut if I knew how)

1. **setting to new value**

    1.1. **in constructor**
        [Socket.cs(165:25):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L165)	_rightEndPoint = new IPEndPoint(
        [Socket.cs(173:25):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L173)	_rightEndPoint = new IPEndPoint(
        [Socket.cs(180:25):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L180)	_rightEndPoint = new UnixDomainSocketEndPoint(IPEndPointExtensions.GetNetSocketAddress(socketAddress));
        [Socket.Windows.cs(69:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Windows.cs#L69)	_rightEndPoint = ep.Create(socketAddress);

    1.2. **after non-blocking connect**
        [LocalEndPoint] [Socket.cs(319:21):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L319)	_rightEndPoint = _nonBlockingConnectRightEndPoint;
        [RemoteEndPoint] [Socket.cs(361:25):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L361)	_rightEndPoint = _nonBlockingConnectRightEndPoint;
        [Connected] [Socket.cs(472:21):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L472)	_rightEndPoint = _nonBlockingConnectRightEndPoint;

    1.3. **if it was null before**
        [DoBind] [Socket.cs(834:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L834)	_rightEndPoint = endPointSnapshot;
        [SendTo] [Socket.cs(1329:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L1329)	_rightEndPoint = remoteEP;
        [ReceiveMessageFrom] [Socket.cs(1566:21):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L1566)	_rightEndPoint = endPointSnapshot;
        [ReceiveFrom] [Socket.cs(1650:21):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L1650)	_rightEndPoint = endPointSnapshot;
        [DoBeginSendTo] [Socket.cs(2694:21):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L2694)	_rightEndPoint = endPointSnapshot;
        [BeginReceiveMessageFrom] [Socket.cs(3059:21):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3059)	_rightEndPoint = remoteEP;
        [DoBeginReceiveFrom] [Socket.cs(3276:21):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3276)	_rightEndPoint = endPointSnapshot;
        [ConnectAsync] [Socket.cs(3677:21):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3677)	_rightEndPoint = endPointSnapshot;
        [SendToAsync] [Socket.cs(4019:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4019)	_rightEndPoint = endPointSnapshot;
        [DoConnect] [Socket.cs(4155:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4155)	_rightEndPoint = endPointSnapshot;
        [BeginConnectEx] [Socket.cs(4536:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4536)	_rightEndPoint = endPointSnapshot;

    1.4. **on inheriting other socket state**
        [UpdateAcceptSocket] [Socket.cs(4784:20):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4784)	socket._rightEndPoint = _rightEndPoint;


2. **setting to previous value - on some error**

    2.1. **on ObjectDisposedException, throwing later**
        [DoBeginSendTo] [Socket.cs(2703:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L2703)	_rightEndPoint = oldEndPoint;
        [BeginReceiveMessageFrom] [Socket.cs(3081:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3081)	_rightEndPoint = oldEndPoint;
        [DoBeginReceiveFrom] [Socket.cs(3285:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3285)	_rightEndPoint = oldEndPoint;

    2.2. **on any exception, throwing later**
        [ConnectAsync] [Socket.cs(3700:21):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3700)	_rightEndPoint = oldEndPoint;
        [BeginConnectEx] [Socket.cs(4547:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4547)	_rightEndPoint = oldEndPoint;

    2.3. **on false CheckErrorAndUpdateStatus, throwing later**
        [DoBeginSendTo] [Socket.cs(2712:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L2712)	_rightEndPoint = oldEndPoint;
        [BeginReceiveMessageFrom] [Socket.cs(3090:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3090)	_rightEndPoint = oldEndPoint;
        [DoBeginReceiveFrom] [Socket.cs(3294:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3294)	_rightEndPoint = oldEndPoint;
        [BeginConnectEx] [Socket.cs(4563:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4563)	_rightEndPoint = oldEndPoint;

    2.4. **on false CheckErrorAndUpdateStatus, NO throwing**
        [SendToAsync] [Socket.cs(4037:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4037)	_rightEndPoint = oldEndPoint;


3. **setting to null - on some error**

    3.1. **on any exception, throwing later**
        [SendToAsync] [Socket.cs(4029:17):](https://github.com/dotnet/runtime/blob/2bfbb56fdc293f9ca44702782bd020805e4fa428/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4029)	_rightEndPoint = null;

You suggest to clear _localEndPoint on the reset of _rightEndPoint, which are items 2. and 3. on the list. If we look at subitems of item 1., we can see that most of them are if _rightEndPoint was null before or in the constructor, which means if we clear _localEndPoint at this time we won't actually change anything. And a change of _rightEndPoint after non-blocking connect will only happen once I believe.

What I am trying to say is if we want to clear _localEndPoint on the reset of _rightEndPoint, maybe we could actually return to my original proposal of clearing _localEndPoint on every _rightEndPoint change? The only place where it might possibly unnesessary clear some cached value is this non-blocking connect part (item 1.2.) and we might use _rightEndPoint there instead of new property RightEndPoint specifically in this case.

CarnaViire committed:
Merge branch 'master' into issue-1482

CarnaViire committed:
Add clear on error

CarnaViire committed:
Add clear on connect and tests

CarnaViire committed:
Add caching test

antonfirsov commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/LocalEndPointTest.cs:
```diff
@@ -0,0 +1,154 @@
+            using (Socket server = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
+            using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
+            {
+                server.Bind(new IPEndPoint(IPAddress.Any, 0));
+
+                Assert.Null(client.LocalEndPoint);
+
+                client.Bind(new IPEndPoint(IPAddress.Any, 0));
+
+                IPEndPoint localEPAfterBind = (IPEndPoint) client.LocalEndPoint;
+                Assert.Equal(IPAddress.Any, localEPAfterBind.Address);
+                int portAfterBind = localEPAfterBind.Port;
+
+                var sendToEP = new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)server.LocalEndPoint).Port);
```
We have the `SocketTestExtensions.BindToAnonymousPort` helper to make this kind of code simpler.

CarnaViire replied:
Fixed

CarnaViire committed:
Use BindToAnonymousPort

CarnaViire committed:
Merge remote-tracking branch 'upstream/master' into issue-1482

CarnaViire committed:
Fix typo

CarnaViire commented:
/azp run runtime-libraries outerloop

CarnaViire committed:
Merge remote-tracking branch 'upstream/master' into issue-1482

CarnaViire committed:
Merge remote-tracking branch 'upstream/master' into issue-1482

CarnaViire commented:
/azp run runtime-libraries outerloop

CarnaViire committed:
Merge remote-tracking branch 'upstream/master' into issue-1482

geoffkizer commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -4848,6 +4871,7 @@ internal Socket UpdateAcceptSocket(Socket socket, EndPoint remoteEP)
             socket._protocolType = _protocolType;
             socket._rightEndPoint = _rightEndPoint;
             socket._remoteEndPoint = remoteEP;
+            socket._localEndPoint = null;
```
It seems like here we could get the local end point from the listener, right?

CarnaViire replied:
If I do
```c#
socket._localEndPoint = _localEndPoint;
```
then some tests fail, for example in `DualModeBeginAccept.BeginAcceptV4BoundToAnyV4_Success` with address being `::ffff:0.0` or so instead of expected `::ffff:127.0.0.1` here https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Sockets/tests/FunctionalTests/DualModeSocketTest.cs#L857
Or you were suggesting something else?

geoffkizer replied:
Yeah, that's what I was suggesting.

How do the tests fail? Is it only dual mode tests that fail?

The code as-is seems fine, but if there's a reason we can't do my suggestion, it would probably be good to add a comment explaining why so that we don't try to "optimize" this in the future and hit the same issue.

CarnaViire replied:
I've tried it now and didn't get these failures as I did before 😞 It seems that one of my later changes fixed it. So I added the assignment. Sorry for misinformation.

CarnaViire replied:
So the test failures are actually back in CI. I will do some more investigation to understand the behavior of sockets there.

antonfirsov replied:
This is really weird, looks like a wildcard address is being set to `_localEndpoint` in the listener socket, which is then copied to the accept socket. If you can get to the bottom of it, it would give us some good insight.

@geoffkizer it looks common sense that the address shall be copied to the accept socket, but ideally I would like to see `getsockname` and `accept` docs guaranteeing this, especially if there may be weird corner cases happening around dual mode.

geoffkizer replied:
I'm fine with that. I mostly just want to make sure we understand this and comment it appropriately in the code. The wildcard binding case seems like good one to understand and comment about in the code.

CarnaViire commented:
/azp run runtime-libraries outerloop

CarnaViire committed:
Assign _localEndPoint from the listener

CarnaViire commented:
/azp run runtime-libraries outerloop

CarnaViire committed:
tmp commit to check on mac

CarnaViire committed:
tmp

CarnaViire committed:
Merge remote-tracking branch 'upstream/master' into issue-1482

CarnaViire committed:
Merge branch 'master' into issue-1482

CarnaViire committed:
Fix _localEndPoint clearing on mac

CarnaViire commented:
So the behavior with wildcard addresses is as follows.

1) If a TCP server socket (listener socket) was bound to a wildcard address, then the accept socket created with `accept` will be assigned with a specific address based on what interface a client socket connected to. That means we can't copy cached listener's local endpoint if it is wildcard. However, if the listener was bound to a specific address, we can copy that to the accept socket freely.

2) If a TCP client socket was bound to a wildcard address, then it will be assigned with a specific address after `connect`. That means we should clear the client socket's cached local endpoint if it is wildcard.

3) If an UDP socket was bound to a wildcard address, it stays as wildcard even after `sendto` and other calls.

I added clearing of cached wildcard local endpoints on `connect` and `accept` with clarifying comments. Also added tests that illustrate this behavior.

cc @geoffkizer @antonfirsov

CarnaViire commented:
/azp run runtime-libraries outerloop

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -22,14 +22,22 @@ public partial class Socket : IDisposable
     {
         internal const int DefaultCloseTimeout = -1; // NOTE: changing this default is a breaking change.
 
+        private static readonly HashSet<IPAddress> s_wildcardAddresses = new HashSet<IPAddress>
```
With only 4 items an array would be cheaper here.

CarnaViire replied:
Fixed

CarnaViire committed:
Inline wildcard addresses

CarnaViire committed:
Merge remote-tracking branch 'upstream/master' into issue-1482

CarnaViire commented:
/azp run runtime-libraries outerloop

CarnaViire commented:
Outerloop test failures are #40926 and #41402

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -215,6 +219,7 @@ private unsafe Socket(SafeSocketHandle handle, bool loadPropertiesFromHandle)
                                 }
 
                                 _isConnected = true;
+                                UpdateLocalEndPointOnConnect();
```
This is a constructor. I can't see any prior code assigning `_localEndpoint` before this call. In that case we don't need this.

CarnaViire replied:
Fixed

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -224,6 +229,7 @@ private unsafe Socket(SafeSocketHandle handle, bool loadPropertiesFromHandle)
                                 // whether we're actually connected or not, err on the side of saying
                                 // we're connected.
                                 _isConnected = true;
+                                UpdateLocalEndPointOnConnect();
```
Same as above.

CarnaViire replied:
Fixed

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -4880,9 +4912,38 @@ internal void SetToConnected()
+            if (endPoint is IPEndPoint ipEndpoint)
+            {
+                IPAddress address = ipEndpoint.Address;
+                return IPAddress.Any.Equals(address) || IPAddress.IPv6Any.Equals(address) || IPAddress.Any.MapToIPv6().Equals(address);
```
This will allocate on all calls to `IsWildcardEndpoint`. We may consider caching `IPAddress.Any.MapToIPv6()` to a `static readonly` variable, and replace all usages.

CarnaViire replied:
Fixed

CarnaViire committed:
PR fixes

antonfirsov commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/LocalEndPointTest.cs:
```diff
@@ -0,0 +1,260 @@
+
+                var sendToEP = new IPEndPoint(Loopback, serverPort);
+
+                client.SendTo(new byte[] { 1, 2, 3 }, sendToEP);
+
+                Assert.Equal(Wildcard, GetLocalEPAddress(client)); // wildcard after sendto
+
+                byte[] buf = new byte[3];
+                EndPoint receiveFromEP = new IPEndPoint(Wildcard, 0);
+                server.ReceiveFrom(buf, ref receiveFromEP);
+
+                Assert.Equal(new byte[] { 1, 2, 3 }, buf);
+            }
+        }
+
+        [Fact]
+        public void UdpSocket_ClientNotBound_LocalEPBecomesWildcardOnAsyncSendTo()
+        {
+            using (Socket server = CreateUdpSocket())
+            using (Socket client = CreateUdpSocket())
+            {
+                int serverPort = server.BindToAnonymousPort(Wildcard);
+
+                Assert.Null(client.LocalEndPoint); // null before async WSASendTo
+
+                var sendToEP = new IPEndPoint(Loopback, serverPort);
+
+                IAsyncResult sendToResult = client.BeginSendTo(new byte[] { 4, 5, 6 }, 0, 3, SocketFlags.None, sendToEP, null, null);
+                sendToResult.AsyncWaitHandle.WaitOne();
+                client.EndSendTo(sendToResult);
+
+                Assert.Equal(Wildcard, GetLocalEPAddress(client)); // wildcard after async WSASendTo
+
+                byte[] buf = new byte[3];
+                EndPoint receiveFromEP = new IPEndPoint(Wildcard, 0);
+                server.ReceiveFrom(buf, ref receiveFromEP);
+
+                Assert.Equal(new byte[] { 4, 5, 6 }, buf);
+            }
+        }
```
These 2 tests seem to test the same thing for two different connect implementations. I suggest dedupe their code and also cover the remaining ones: `ConnectAsync(SocketAsyncEventArgs)`  and `SocketTaskExtensions.ConnectAsync(this Socket, ....)`. For this the best would be to utilize  `SocketTestHelperBase<T>`.

CarnaViire replied:
Rewrote the tests, please check that I understood you correctly :)

antonfirsov reviewed (approved):
We need to make sure all independent implementations are covered in the tests, otherwise looks good now.

CarnaViire committed:
Rewrite tests with SocketTestHelperBase
